### ISSUE TITLE
Reconsider Debug.Listeners API exposure

### ISSUE THREAD

**Issue by mwpowellhtx** [dotnet/corefx#40167]
I am staring at a use case right now that I need to listen for ``System.Diagnostics.Debug.WriteLine`` messages and a listener would be just what I need to get the job done. Therefore, I strongly suggest [reconsidering Debug.Listeners exposure](/dotnet/standard/issues/84). Absent that we must look at adding unnecessary layers of overhead to our ``Debug.WriteLine`` listening efforts when a listener would fit the bill nicely.

**Comment by terrajobst** [comment]
I've moved this to CoreFX, because we first would have to make that change here.

@noahfalk, @jkotas @stephentoub are the reasons to not expose `Debug.Listeners` still sensible today? As far as I can see, `Debug` lives in `System.Private.CorLib` while `TraceListener` lives in `System.Diagnostics.TraceSource`. We'd have to move this to CorLib.

**Comment by mwpowellhtx** [comment]
I've got a workaround in my library that measures when the API sees a message occur, but this is not a substitute for having a *Debug Listener*, which I need because the 3P library I am building upon routes messages through ``Debug.WriteLine``. This is what I really need to measure, not not just to mock anything. I want to measure with hard data, an actual message. What that means for internal sensibilities, cannot say, don't care. What I've got now is a poor, surface, substitute for what I think is the long term solution.

**Comment by stephentoub** [comment]
To make sure it's known, Trace.Listeners in .NET Core 3.0 will serve as Debug listeners.

**Comment by jkotas** [comment]
> We'd have to move this to CoreLib.

Including its dependency closure. The dependency closure was the reason why we have not done this originally. I guess it may not be as bad as it used to be given that we have moved a ton of other stuff to CoreLib. It just makes the reasoning about layering and treeshake-ability harder.

> Trace.Listeners in .NET Core 3.0 will serve as Debug listeners.

Yes, this was fixed https://github.com/dotnet/corefx/commit/350f100edbba9e88aad3a9b030b401b2bea0469b . `Trace.Listeners` in .NET Core 3.0 is functionally equivalent to what `Debug.Listeners` would do.

**Comment by mwpowellhtx** [comment]
@jkotas Would you mind clarifying, are you saying that if we install a listener in the ``Trace.Listeners``, messages routing through ``Debug.Whatever`` will be seen there?

**Comment by stephentoub** [comment]
Yes.  This:
```C#
using System;
using System.Diagnostics;

class Program
{
    static void Main()
    {
        Trace.Listeners.Clear();
        Trace.Listeners.Add(new ConsoleTraceListener());
        Debug.Fail("hello");
        Console.WriteLine("got here");
    }
}
```
on .NET Core 3.0 outputs:
```
Fail: hello
got here
```

**Comment by mwpowellhtx** [comment]
@stephentoub To be clear, we are not talking about monitoring ``Debug.Fail``. But rather, ``Debug.WriteLine``. We do not care about ``Console.WriteLine`` for this purpose; our 3P subscription uses ``Debug.WriteLine``, beyond our immediate control, so we want to monitor for that when our wrapper invoked the 3P API.

**Comment by stephentoub** [comment]
>To be clear, we are not talking about monitoring Debug.Fail. But rather, Debug.WriteLine

Yes, that was just an example.  All of these APIs go through the same code paths.  The following:
```C#
using System;
using System.Diagnostics;

class Program
{
    static void Main()
    {
        Trace.Listeners.Clear();
        Trace.Listeners.Add(new ConsoleTraceListener());
        Debug.WriteLine("hello");
        Console.WriteLine("got here");
    }
}
```
prints:
```
hello
got here
```
